c-dependencies/js-compute-runtime/js-compute-builtins.cpp

```diff
@@ -5451,6 +5474,437 @@ bool constructor(JSContext *cx, unsigned argc, Value *vp) {
+
+ObjectStoreHandle object_store_handle(JSObject *obj) {
+  JS::Value val = JS::GetReservedSlot(obj, Slots::ObjectStore);
+  return ObjectStoreHandle{static_cast<uint32_t>(val.toInt32())};
+}
+
+const unsigned ctor_length = 1;
+
+bool check_receiver(JSContext *cx, HandleValue receiver, const char *method_name);
+
+bool lookup(JSContext *cx, unsigned argc, Value *vp) {
+  METHOD_HEADER(1)
+
+  RootedObject result_promise(cx, JS::NewPromiseObject(cx, nullptr));
+  if (!result_promise) {
+    return ReturnPromiseRejectedWithPendingError(cx, args);
+  }
+
+  size_t key_len;
+  // Convert the key argument into a String following https://tc39.es/ecma262/#sec-tostring
+  UniqueChars key = encode(cx, args.get(0), &key_len);
+  if (!key) {
+    return false;
+  }
+
+  // If the converted string has a length of 0 then we throw an Error
+  // because ObjectStore Keys have to be at-least 1 character.
+  // TODO: Decide if this should be a RangeError
+  if (key_len == 0) {
+    JS_ReportErrorUTF8(cx, "ObjectStore key can not be an empty string");
+    return false;
+  }
+
+  // If the converted string has a length of more than 1024 then we throw an Error
+  // because ObjectStore Keys have to be less than 1025 characters.
+  // TODO: Decide if this should be a RangeError
+  if (key_len > 1024) {
+    JS_ReportErrorUTF8(cx, "ObjectStore key can not be more than 1024 characters");
+    return false;
+  }
+
+  char *key_chars = key.get();
+
+  if (strchr(key_chars, '#') != NULL) {
+    JS_ReportErrorASCII(cx, "ObjectStore key can not contain # character");
+    return false;
+  }
+  if (strchr(key_chars, '?') != NULL) {
+    JS_ReportErrorASCII(cx, "ObjectStore key can not contain ? character");
+    return false;
+  }
+  if (strchr(key_chars, '*') != NULL) {
+    JS_ReportErrorASCII(cx, "ObjectStore key can not contain * character");
+    return false;
+  }
+  if (strchr(key_chars, '[') != NULL) {
+    JS_ReportErrorASCII(cx, "ObjectStore key can not contain [ character");
+    return false;
+  }
+  if (strchr(key_chars, ']') != NULL) {
+    JS_ReportErrorASCII(cx, "ObjectStore key can not contain ] character");
+    return false;
+  }
+  if (strchr(key_chars, '\n') != NULL) {
+    JS_ReportErrorASCII(cx, "ObjectStore key can not contain newline character");
+    return false;
+  }
+  if (strchr(key_chars, '\r') != NULL) {
+    JS_ReportErrorASCII(cx, "ObjectStore key can not contain carriage return character");
+    return false;
+  }
+  auto acme_challenge = ".well-known/acme-challenge/";
+  if (strncmp(key_chars, acme_challenge, strlen(acme_challenge)) == 0) {
+    JS_ReportErrorASCII(cx, "ObjectStore key can not start with .well-known/acme-challenge/");
+    return false;
+  }
+
+  if (strcmp(key_chars, ".") == 0) {
+    JS_ReportErrorASCII(cx, "ObjectStore key can not be \".\"");
+    return false;
+  }
+  if (strcmp(key_chars, "..") == 0) {
+    JS_ReportErrorASCII(cx, "ObjectStore key can not be \"..\"");
+    return false;
+  }
```

Comment:
These checks exist within the ObjectStore implementation at https://github.com/fastly/object-store
The reason we have them implemented in the JS runtime is to be able to throw a JS Errors that have really nice error message for each scenario instead of the generic "Invalid argument. - Fastly error code 2" message that HANDLE_RESULT would create